### PR TITLE
Add const generic params

### PR DESCRIPTION
**Pull request by oli-obk**
Best reviewed commit by commit.

Before I continue down this road, does this work look in the spirit of how you're imagining a-mir-formality to go?

### CODE REVIEW

File: crates/formality-types/src/grammar/ty.rs [hunk 1]
```diff
@@ -113,7 +117,7 @@ impl UpcastFrom<Ty> for TyData {
     }
 }
 
-#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
```
Comment (nikomatsakis) [inline crates/formality-types/src/grammar/ty.rs:123]: Why make Variable `!Copy`?

**Review by nikomatsakis — APPROVED** [review]
This looks generally right but I'm not sure why you are making `Variable: !Copy`.

**Comment by nikomatsakis** [comment]
In particular, I have some thoughts about how to handle specifying the type of constants. I think it wants to be stored in the environment as a predicate.

**Comment by oli-obk** [comment]
Yea, I have recently (Friday) looked into removing the type of constants from them, and if we could first show that this works in a-mir-formality, that would be great. I'll investigate if I can do this PR with the predicate instead, but that has its own problems (that are easier to investigate in a-mir-formality than in rustc)

**Comment by oli-obk** [comment]
> In particular, I have some thoughts about how to handle specifying the type of constants. I think it wants to be stored in the environment as a predicate. (I am guessing this may be related to my question above.)

I did this in https://github.com/rust-lang/a-mir-formality/pull/130

We still need to store a type for rigid constants, but we don't need the type in ParameterKind anymore.

**Pull request closed without merging**